internal/util/get/get.go

```diff
@@ -67,41 +68,142 @@ func (c Command) Run() error {
 	// define where we are going to clone the package from
 	r := &git.RepoSpec{OrgRepo: c.Repo, Path: c.Directory, Ref: c.Ref}
 
-	defaultRef, err := gitutil.DefaultRef(c.Repo)
+	err := cloneAndCopy(r, c.Destination, c.Name, c.Clean)
 	if err != nil {
 		return err
 	}
 
-	// clone the repo to a tmp directory.
-	// delete the tmp directory later.
-	err = ClonerUsingGitExec(r, defaultRef)
+	if err = c.fetchRemoteSubpackages(); err != nil {
+		return errors.Wrap(err)
+	}
+	return nil
+}
+
+// fetchRemoteSubpackages goes through the root package and its subpackages
+// and fetches any remote subpackages referenced. It will also handle situations
+// where a remote subpackage references other remote subpackages.
+func (c Command) fetchRemoteSubpackages() error {
+	// Create a stack to keep track of all Kptfiles that needs to be checked
+	// for remote subpackages.
+	stack := newStack()
+
+	paths, err := pathutil.DirsWithFile(c.Destination, kptfile.KptFileName, true)
+	if err != nil {
+		return err
+	}
+	for _, p := range paths {
+		stack.push(p)
+	}
+
+	for stack.len() > 0 {
+		p := stack.pop()
+		kf, err := kptfileutil.ReadFile(p)
+		if err != nil {
+			return err
+		}
+
+		remoteSubPkgDirs := make(map[string]bool)
+		for i := range kf.Subpackages {
+			sp := kf.Subpackages[i]
+
+			if _, found := remoteSubPkgDirs[sp.LocalDir]; found {
+				return fmt.Errorf("multiple remote subpackages with localDir %q", sp.LocalDir)
+			}
+			remoteSubPkgDirs[sp.LocalDir] = true
+
+			gitInfo := sp.Git
+			localPath := filepath.Join(p, sp.LocalDir)
+
+			_, err = os.Stat(localPath)
+			if err != nil && !os.IsNotExist(err) {
+				return err
+			}
+			if !os.IsNotExist(err) {
+				return fmt.Errorf("local subpackage in directory %q already exist", sp.LocalDir)
+			}
+
+			r := &git.RepoSpec{OrgRepo: gitInfo.Repo, Path: gitInfo.Directory, Ref: gitInfo.Ref}
+			err := cloneAndCopy(r, localPath, sp.LocalDir, false)
+			if err != nil {
+				return err
+			}
+
+			subPaths, err := pathutil.DirsWithFile(localPath, kptfile.KptFileName, true)
+			if err != nil {
+				return err
+			}
+			for _, subp := range subPaths {
+				if subp == localPath {
+					continue
+				}
+				stack.push(subp)
+			}
+		}
+	}
+	return nil
+}
+
+// cloneAndCopy fetches the provided repo and copies the content into the
+// directory specified by dest. The provided name is set as `metadata.name`
+// of the Kptfile of the package.
+func cloneAndCopy(r *git.RepoSpec, dest, name string, clean bool) error {
```

Comment:
Does this also add a flag in the Kptfile indicating if this is fetched as a remote subpackage in order to differentiate it from the local subpackages?

Reply:
No, it doesn't. I haven't seen a need for that yet. We can consider adding it if we need it.